### PR TITLE
fix: v2sku for APIM - `avm/res/api-management/service`

### LINKED ISSUE #1473: [WAF - Reliability Review] - `avm/res/api-management/service`
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

avm/res/api-management/service

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

Hello team,

The AVM core team are performing a review of all AVM modules to ensure they are aligned to the WAF reliability pillar by default, and in the future they will be checking against other WAF pillars also. This is a requirement we have before we can go `v1.0.0` for any AVM module.

Links to AVM specs:

- https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-sfr2---category-composition---waf-aligned
- https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-sfr5---category-composition---availability-zones

Links to WAF Reliability, APRL & Product Docs recommendations used for analysis:

- https://learn.microsoft.com/azure/well-architected/reliability/checklist
- https://learn.microsoft.com/en-us/azure/api-management/high-availability?toc=%2Fazure%2Freliability%2Ftoc.json&bc=%2Fazure%2Freliability%2Fbreadcrumb%2Ftoc.json
- https://azure.github.io/Azure-Proactive-Resiliency-Library/services/integration/api-management

Whilst reviewing this module the following issues have been found that are non-compliant with the AVM specs that need to be resolved in a new version release:

- [Change the default for the `skuCount` parameter to `3` (to match the zones) and adjust/remove the allowed decorator for the module](https://github.com/Azure/bicep-registry-modules/blob/main/avm/res/api-management/service/main.bicep#L68-L71)
- [Change the default for the `zones` parameter to `[1,2,3]`](https://github.com/Azure/bicep-registry-modules/blob/main/avm/res/api-management/service/main.bicep#L92)
- [Change the default for the `sku` parameter to `Premium` tier to support AZs](https://github.com/Azure/bicep-registry-modules/blob/main/avm/res/api-management/service/main.bicep#L65)

Please make these changes in a PR to the parameter defaults so a new minor version can be published.

### PSRule

It also seems that PSRule is not picking up these checks under its [`Azure.Pillar.Reliability` baseline](https://azure.github.io/PSRule.Rules.Azure/en/baselines/Azure.Pillar.Reliability/) via its tests and that has been logged in the issue: https://github.com/Azure/PSRule.Rules.Azure/issues/2787

And there is also a missing check for the capacity instances which has been logged here: https://github.com/Azure/PSRule.Rules.Azure/issues/2788

Thanks

The AVM Core Team

### (Optional) Correlation Id

_No response_

### LINKED ISSUE #1929: [AVM Module Issue]: api management v2 sku not working due to api version
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Bug

### Module Name

avm/res/api-management/service

### (Optional) Module Version

0.1.7

### Description

Hi team,

Im using the AVM module /res/api-management/service:0.1.7 (latest version) which supports the new V2 sku's.
According to this pull request this should be supported now: https://github.com/Azure/bicep-registry-modules/pull/1777

During deployment I got an error (Bad Request) which the message:   "Operation on StandardV2 SKU is only supported in the api-versions 2023-03-01-preview,2023-05-01-preview,2023-09-01-preview (Code: NotSupported)"
image

image

This is my Bicep configuration:

```bicep
module modApiManagementService 'br/public:avm/res/api-management/service:0.1.7' = {
  scope: 'resourcegroup'
  name: '${uniqueString(deployment().name)}-apim'
  params: {
    name: 'apinname'
    publisherEmail: 'email@test.nl'
    publisherName: 'somename'
    location: 'location'
    sku: 'StandardV2'
    virtualNetworkType:'Internal'
    subnetResourceId: 'somesubnet'
  }
}
```

### (Optional) Correlation Id

_No response_

### LINKED ISSUE #2075: [AVM Module Issue]: Urgent action required: Update your Azure API Management tooling to a supported management API version by 1 June 2024
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

avm/res/api-management/service

### (Optional) Module Version

_No response_

### Description

Hi,

After deployment of APIM with diagnostic settings, we get a message in Azure Health Advisory:

> You're receiving this message because you're currently using one or more API versions prior to 2021-08-01 to configure or manage your Azure API Management services.
> 
> All API versions prior to 2021-08-01 will be retired starting 1 June 2024. These API versions will be disabled gradually and without further notice. The previously communicated date was 30 September 2023. You'll no longer be able to create or manage your API Management services using templates, tools, scripts, or programs that rely on a disabled API version.
> 
> Data plane operations, such as mediation of API requests in the gateway, remain unaffected by this change.
> 
> The API version may be embedded in deployment templates and tools that interact programmatically with API Management services.

More info: https://learn.microsoft.com/en-us/azure/api-management/breaking-changes/api-version-retirement-sep-2023

The main.bicep config file contains the resource diagnostic settings with an API version prio to api-version 2021-08-01 which triggers the advisory alert.

https://github.com/Azure/bicep-registry-modules/blob/67c4156b2fec8c18116c2fe15791931708716666/avm/res/api-management/service/main.bicep#L468

Can we take this opportunity to upgrade all apim resources in this module to a more recent api-version ? ex: 2022-08-01

### (Optional) Correlation Id

_No response_

### PR DESCRIPTION
**Pull request by tony-box**
## Description

Closes #1929 
Closes #2075 
Closes #1473
- Update all api-management/service/* modules to use newer API versions (2022-08-01)
- Update api-management/service api version 2023-05-01-preview to allow v2sku options
- Add V2sku deployment option for api-management/service
- adds an additional test for v2 sku deployment.
- modify all tests to deploy APIM with Premium SKU, and with multiple zones and 2nd region to appease latest PS Rule tests.

## Pipeline Reference

| Pipeline |
| -------- |
|    [avm.res.api-management.service](https://github.com/tony-box/bicep-registry-modules/actions/workflows/avm.res.api-management.service.yml)      |

## Type of Change

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [x] Azure Verified Module updates:
  - [x] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [x] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [x] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/avm_module_issue.yml [hunk 1]
```diff
@@ -45,7 +45,6 @@ body:
         # - "avm/ptn/avd-lza/management-plane"
         # - "avm/ptn/avd-lza/networking"
         # - "avm/ptn/avd-lza/session-hosts"
-        - "avm/ptn/policy-insights/remediation"
```
Comment (eriqua) [inline .github/ISSUE_TEMPLATE/avm_module_issue.yml:48]: Is this removal intended?

**Comment by eriqua** [comment]
@tony-box please also have a look at #1795 which is also targeting the same resource module and may thus result in conflicts to be addressed. 
I'd suggest to prioritize addressing comments there first

File: .github/ISSUE_TEMPLATE/avm_module_issue.yml (hunk 1 above)
Comment (tony-box) [reply .github/ISSUE_TEMPLATE/avm_module_issue.yml]: @eriqua I fixed this. my fork was messed up and had some weird commits in it that were getting dragged along into all my other branches. This should be a clean commit now with just the changes related to the fix.

**Comment by tony-box** [comment]
> @tony-box please also have a look at #1795 which is also targeting the same resource module and may thus result in conflicts to be addressed. I'd suggest to prioritize addressing comments there first

There shouldn't be any conflicts between these two PRs. Regardless, I pretty much have to redo that whole PR anyway due to all the weird commits that were included in it (similar to this one). I will be retesting it and ensuring there are no issues before that one goes through.

**Comment by ChrisSidebotham** [comment]
Converting back to draft until @tony-box is ready :)

**Comment by tony-box** [comment]
I'm going to update the rest of the APIM services' API version in this PR to fix #2075 as well.

File: avm/res/api-management/service/tests/e2e/defaults/main.test.bicep [hunk 2]
```diff
@@ -45,6 +45,19 @@ module testDeployment '../../../main.bicep' = [
...
+      additionalLocations: [
+        {
+          location: 'westus'
+          sku: {
+            name: 'Premium'
+            capacity: 1
+          }
+          disableGateway: false
+        }
+      ]
```
Comment (AlexanderSehr) [inline avm/res/api-management/service/tests/e2e/defaults/main.test.bicep:60]: Are those parameters required?

**Comment by AlexanderSehr** [comment]
Hey @tony-box, I see there was some back and forth regarding this PR. Would you say it's ready?

**Comment by tony-box** [comment]
closing this PR which will be superceeded by #2453
